bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java

```diff
@@ -123,261 +116,59 @@
-        }
-
-        @Override
-        public void statusUpdated(Thing thing, ThingStatusInfo statusInfo) {
-            // note: all provoked operations based on a status update should be executed asynchronously!
-            ThingStatusInfo oldStatusInfo = thing.getStatusInfo();
-            ensureValidStatus(oldStatusInfo.getStatus(), statusInfo.getStatus());
-
-            if (ThingStatus.REMOVING.equals(oldStatusInfo.getStatus())
-                    && !ThingStatus.REMOVED.equals(statusInfo.getStatus())) {
-                // if we go to ONLINE and are still in REMOVING, notify handler about required removal
-                if (ThingStatus.ONLINE.equals(statusInfo.getStatus())) {
-                    logger.debug(
-                            "Handler is initialized now and we try to remove it, because it is in REMOVING state.");
-                    notifyThingHandlerAboutRemoval(thing);
-                }
-                // only allow REMOVING -> REMOVED transition, all others are ignored because they are illegal
-                logger.debug(
-                        "Ignoring illegal status transition for thing {} from REMOVING to {}, only REMOVED would have been allowed.",
-                        thing.getUID(), statusInfo.getStatus());
-                return;
-            }
-
-            // update thing status and send event about new status
-            setThingStatus(thing, statusInfo);
-
-            // if thing is a bridge
-            if (isBridge(thing)) {
-                handleBridgeStatusUpdate((Bridge) thing, statusInfo, oldStatusInfo);
-            }
-            // if thing has a bridge
-            if (hasBridge(thing)) {
-                handleBridgeChildStatusUpdate(thing, oldStatusInfo);
-            }
-            // notify thing registry about thing removal
-            if (ThingStatus.REMOVED.equals(thing.getStatus())) {
-                notifyRegistryAboutForceRemove(thing);
-            }
-        }
-
-        private void ensureValidStatus(ThingStatus oldStatus, ThingStatus newStatus) {
-            if (!(ThingStatus.UNKNOWN.equals(newStatus) || ThingStatus.ONLINE.equals(newStatus)
-                    || ThingStatus.OFFLINE.equals(newStatus) || ThingStatus.REMOVED.equals(newStatus))) {
-                throw new IllegalArgumentException(MessageFormat.format(
-                        "Illegal status {0}. Bindings only may set {1}, {2}, {3} or {4}.", newStatus,
-                        ThingStatus.UNKNOWN, ThingStatus.ONLINE, ThingStatus.OFFLINE, ThingStatus.REMOVED));
-            }
-            if (ThingStatus.REMOVED.equals(newStatus) && !ThingStatus.REMOVING.equals(oldStatus)) {
-                throw new IllegalArgumentException(
-                        MessageFormat.format("Illegal status {0}. The thing was in state {1} and not in {2}", newStatus,
-                                oldStatus, ThingStatus.REMOVING));
-            }
-        }
```

Comment:
@jlaur, this code was already there, I just moved it from an anonymous class in `ThingHandlerImpl` to a separate class. If it didn't work, that would be a bug.

Reply:
You are right. I will test later what the actual current behavior is. I now also noticed this slight inconsistency which seems to have been made on purpose: 6258268031ecc4dedc832f59de01152686fc3cf8. Not sure why this invalid transition would need to be handled "softer" than others.

Reply:
I finally got around to test this. They are already broken today:
```
java.lang.IllegalArgumentException: Illegal status UNINITIALIZED. Bindings only may set UNKNOWN, ONLINE, OFFLINE or REMOVED.
	at org.openhab.core.thing.internal.ThingManagerImpl$1.ensureValidStatus(ThingManagerImpl.java:230) ~[?:?]
	at org.openhab.core.thing.internal.ThingManagerImpl$1.statusUpdated(ThingManagerImpl.java:193) ~[?:?]
	at org.openhab.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:380) ~[?:?]
	at org.openhab.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:404) ~[?:?]
```